### PR TITLE
CLI Binding for String Encoding

### PR DESCRIPTION
**Pull request by jeffin143**
@lozhnikov , Added binding for String Encoding and also added respective test for the binding. please take a look.

### CODE REVIEW

**Comment by lozhnikov** [comment]
I'll take a look today.

**Comment by lozhnikov** [comment]
@jeffin143 I looked through the code. I think these bindings share a lot of code with the string cleaning bindings. I guess we could move the shared code into a separate file.

File: src/mlpack/methods/preprocess/preprocess_string_util.hpp [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+
+/**
+ * Function neccessary to create a vector<vector<string>> by readin
+ * the contents of a file.
+ *
+ * @param filename Name of the file whose contents need to be preproccessed.
+ * @param columnDelimiter Delimiter used to split the columns of file.
+ */
+static vector<vector<string>> CreateDataset(const string& filename,
+                                            char columnDelimiter)
```
Comment (lozhnikov) [inline src/mlpack/methods/preprocess/preprocess_string_util.hpp]: It's not a template function. There is no need to write the implementation in a header file. I suggest to move the implementation to a source code file.

File: src/mlpack/methods/preprocess/preprocess_string_encoding_main.cpp [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+ *
+ * A CLI executable to encode string dataset.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+
+#include "mlpack/methods/preprocess/preprocess_string_util.cpp"
```
Comment (lozhnikov) [inline src/mlpack/methods/preprocess/preprocess_string_encoding_main.cpp]: `.cpp` files are not supposed to be included.

File: src/mlpack/methods/preprocess/preprocess_string_encoding_main.cpp [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+/**
+ * Function used to write back the preproccessed data to a file.
+ *
+ * @param outputFilename Name of the file to save the data into.
+ * @param dataset The actual dataset which was read from file.
+ * @param columnDelimiter Delimiter used to separate columns of the file.
+ * @param dimesnions Dimesnion which we non numeric or of type string.
+ * @param encodedResult Collection of arma matrices havinf encoded Results.
+ */
+static void WriteOutput(const string& outputFilename,
```
Comment (lozhnikov) [inline src/mlpack/methods/preprocess/preprocess_string_encoding_main.cpp]: Why didn't you move this function to a separate file?

File: src/mlpack/methods/preprocess/preprocess_string_util.cpp [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+using namespace std;
+
+/**
+ * Function neccessary to create a vector<vector<string>> by readin
+ * the contents of a file.
+ *
+ * @param filename Name of the file whose contents need to be preproccessed.
+ * @param columnDelimiter Delimiter used to split the columns of file.
+ */
+static vector<vector<string>> CreateDataset(const string& filename,
```
Comment (lozhnikov) [inline src/mlpack/methods/preprocess/preprocess_string_util.cpp]: These functions are used in other files. They shouldn't be static. And I think it's a good idea to move them to a separate namespace in order to avoid conflicts.

File: src/mlpack/methods/preprocess/preprocess_string_encoding_main.cpp (hunk 3 above)
Comment (jeffin143) [reply src/mlpack/methods/preprocess/preprocess_string_encoding_main.cpp]: It is not same in both binding, That's why.

File: src/mlpack/methods/preprocess/preprocess_string_util.hpp [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+
+#ifndef MLPACK_CORE_DATA_STRING_UTIL_HPP
+#define MLPACK_CORE_DATA_STRING_UTIL_HPP
+
+#include <mlpack/prereqs.hpp>
+#include <mlpack/core/util/mlpack_main.hpp>
```
Comment (lozhnikov) [inline src/mlpack/methods/preprocess/preprocess_string_util.hpp]: `mlpack/core/util/mlpack_main.hpp` should only be included in CLI main files. Remove the line, it should fix the error. Also remove redundant includes.

File: src/mlpack/methods/preprocess/preprocess_string_util.hpp [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+#include <mlpack/core/data/string_encoding.hpp>
+#include <mlpack/core/data/tokenizers/split_by_any_of.hpp>
+#include <mlpack/core/data/string_encoding_policies/dictionary_encoding_policy.hpp>
+#include <mlpack/core/data/string_encoding_policies/bag_of_words_encoding_policy.hpp>
+#include <mlpack/core/data/string_encoding_policies/tf_idf_encoding_policy.hpp>
+
+using namespace mlpack;
+using namespace mlpack::util;
+using namespace arma;
+using namespace std;
```
Comment (lozhnikov) [inline src/mlpack/methods/preprocess/preprocess_string_util.hpp]: We should never write `using namespace` in header files.

File: src/mlpack/methods/preprocess/preprocess_string_encoding_main.cpp [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+  else
+    columnDelimiter = data::ColumnDelimiterType(filename);
+
+  // Handling Dimension vector
+  vector<string> tempDimension =
+      CLI::GetParam<vector<string> >("dimension");
+  unordered_set<size_t> dimensions = data::GetColumnIndices(tempDimension);
+  vector<vector<string>> dataset = data::CreateDataset(filename,
+      columnDelimiter[0]);
+  for (auto colIndex : dimensions)
```
Comment (lozhnikov) [inline src/mlpack/methods/preprocess/preprocess_string_encoding_main.cpp]: I suggest to add braces here in order to improve the readability.

File: src/mlpack/methods/preprocess/preprocess_string_encoding_main.cpp [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+    if (colIndex >= dataset.back().size())
+      Log::Fatal << "The index given is out of range, please verify" << endl;
+
+  // Preparing the input dataset on which string manipulation has to be done.
+  // vector<vector<string>> nonNumericInput(dimension.size());
+  unordered_map<size_t , vector<string>> nonNumericInput;
+
+  for (size_t i = 0; i < dataset.size(); i++)
```
Comment (lozhnikov) [inline src/mlpack/methods/preprocess/preprocess_string_encoding_main.cpp]: Again, braces.

File: src/mlpack/methods/preprocess/preprocess_string_encoding_main.cpp [hunk 9]
```diff
@@ -0,0 +1,233 @@
...
+            { return et == "RawCount" || et == "Binary" || et == "SublinearTf"
+            || et == "TermFrequency"; }, true, "Tf Idf encoding type should be "
+            " either RawCount, Binary, SublinearTf or TermFrequency ");
+      }
+      const string tfidfEncodingType =
+          CLI::GetParam<string>("tfidf_encoding_type");
+      if ("RawCount" == tfidfEncodingType)
+      {
+          data::TfIdfEncoding<data::SplitByAnyOf::TokenType>
+            encoder(data::TfIdfEncodingPolicy::TfTypes::RAW_COUNT, !smoothIdf);
```
Comment (lozhnikov) [inline src/mlpack/methods/preprocess/preprocess_string_encoding_main.cpp]: I think it is useful to declare an alias:
```
using TfTypes = data::TfIdfEncodingPolicy::TfTypes;
```

File: src/mlpack/methods/preprocess/preprocess_string_encoding_main.cpp [hunk 10]
```diff
@@ -0,0 +1,233 @@
...
+          encoder.Encode(column.second, output, tokenizer);
+          encodedResult[column.first] = std::move(output);
+      }
+      else if ("Binary" == tfidfEncodingType)
+      {
+        data::TfIdfEncoding<data::SplitByAnyOf::TokenType>
```
Comment (lozhnikov) [inline src/mlpack/methods/preprocess/preprocess_string_encoding_main.cpp]: The same comment:
```
`using TokenType = data::SplitByAnyOf::TokenType;
```

File: src/mlpack/methods/preprocess/preprocess_string_encoding_main.cpp [hunk 11]
```diff
@@ -0,0 +1,233 @@
...
+  // Preparing the input dataset on which string manipulation has to be done.
+  // vector<vector<string>> nonNumericInput(dimension.size());
+  unordered_map<size_t , vector<string>> nonNumericInput;
+
+  for (size_t i = 0; i < dataset.size(); i++)
+    for (auto& datasetCol : dimensions)
+      nonNumericInput[datasetCol].push_back(move(dataset[i][datasetCol]));
+
+  unordered_map<size_t , arma::mat> encodedResult;
+  const string delimiter = CLI::GetParam<string> ("delimiter");
```
Comment (lozhnikov) [inline src/mlpack/methods/preprocess/preprocess_string_encoding_main.cpp]: I think the plural form suits better than the singular one. (`delimiters`).

File: src/mlpack/methods/preprocess/preprocess_string_encoding_main.cpp [hunk 12]
```diff
@@ -0,0 +1,233 @@
...
+  data::SplitByAnyOf tokenizer(delimiter);
```
Comment (lozhnikov) [inline src/mlpack/methods/preprocess/preprocess_string_encoding_main.cpp]: Why don't you support the second tokenizer?

File: src/mlpack/methods/preprocess/preprocess_string_util.hpp [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+ * A CLI executable to encode string dataset.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+
+#ifndef MLPACK_CORE_DATA_STRING_UTIL_HPP
+#define MLPACK_CORE_DATA_STRING_UTIL_HPP
```
Comment (lozhnikov) [inline src/mlpack/methods/preprocess/preprocess_string_util.hpp]: The name has to be `MLPACK_METHODS_PREPROCESS_PREPROCESS_STRING_UTIL_HPP`

File: src/mlpack/methods/preprocess/preprocess_string_util.hpp [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+
+#include <mlpack/prereqs.hpp>
+#include <mlpack/core/math/random.hpp>
```
Comment (lozhnikov) [inline src/mlpack/methods/preprocess/preprocess_string_util.hpp]: You don't need all the includes below, right? Why did you add them here?

File: src/mlpack/methods/preprocess/preprocess_string_util_impl.cpp [hunk 15]
```diff
@@ -0,0 +1,155 @@
+/**
+ * @file preprocess_string_util_impl.hpp
```
Comment (lozhnikov) [inline src/mlpack/methods/preprocess/preprocess_string_util_impl.cpp]: `_impl` isn't needed. I suggest `preprocess_string_util.cpp`.

File: src/mlpack/methods/preprocess/CMakeLists.txt [hunk 16]
```diff
@@ -3,6 +3,8 @@
 set(SOURCES
 scaling_model.hpp
 scaling_model_impl.hpp
+preprocess_string_util.hpp
+preprocess_string_util_impl.cpp
```
Comment (lozhnikov) [inline src/mlpack/methods/preprocess/CMakeLists.txt]: `_impl` isn't needed. I suggest `preprocess_string_util.cpp`.

File: src/mlpack/methods/preprocess/preprocess_string_util_impl.cpp [hunk 17]
```diff
@@ -0,0 +1,155 @@
...
+
+/**
+ * Function neccessary to create a vector<vector<string>> by readin
+ * the contents of a file.
+ *
+ * @param filename Name of the file whose contents need to be preproccessed.
+ * @param columnDelimiter Delimiter used to split the columns of file.
+ */
+vector<vector<string>> CreateDataset(const string& filename,
+                                            char columnDelimiter)
```
Comment (lozhnikov) [inline src/mlpack/methods/preprocess/preprocess_string_util_impl.cpp]: Could you fix the indent?

File: src/mlpack/methods/preprocess/preprocess_string_util_impl.cpp [hunk 18]
```diff
@@ -0,0 +1,155 @@
...
+{
+  vector<vector<string>> dataset;
+  // Extracting the Contents of file
```
Comment (lozhnikov) [inline src/mlpack/methods/preprocess/preprocess_string_util_impl.cpp]: Why does `contents` begin with a capital? No stop at the end of the comment.

File: src/mlpack/methods/preprocess/preprocess_string_util_impl.cpp [hunk 19]
```diff
@@ -0,0 +1,155 @@
...
+}
+
+/**
+ * Function to check whether the given colum has only digits.
+ *
+ * @param column The column no
+ */
+bool IsNumber(const string& column)
+{
+  for (auto i : column)
```
Comment (lozhnikov) [inline src/mlpack/methods/preprocess/preprocess_string_util_impl.cpp]: Again, braces.

File: src/mlpack/methods/preprocess/preprocess_string_util_impl.cpp [hunk 20]
```diff
@@ -0,0 +1,155 @@
...
+  return true;
+}
+
+/**
+ * Function used to get the columns which has non numeric dataset.
+ *
+ * @param tempDimesnion A vector of string passed which has column number or
+ *    column ranges.
+ */
+unordered_set<size_t> GetColumnIndices(const
```
Comment (lozhnikov) [inline src/mlpack/methods/preprocess/preprocess_string_util_impl.cpp]: The line break here looks terrible.

File: src/mlpack/core/data/string_encoding_policies/bag_of_words_encoding_policy.hpp [hunk 21]
```diff
@@ -0,0 +1,141 @@
...
+  /**
+  * The function initializes the output matrix.
+  *
+  * @param output Output matrix to store the encoded results (sp_mat or mat).
+  * @param datasetSize The number of strings in the input dataset.
+  * @param maxNumTokens The maximum number of tokens in the strings of the 
+                        input dataset (not used).
+  * @param dictionarySize The size of the dictionary.
+  * @tparam MatType The type of output matrix.
+  */
```
Comment (zoq) [inline src/mlpack/core/data/string_encoding_policies/bag_of_words_encoding_policy.hpp]:
```suggestion
  /**
   * The function initializes the output matrix.
   *
   * @param output Output matrix to store the encoded results (sp_mat or mat).
   * @param datasetSize The number of strings in the input dataset.
   * @param maxNumTokens The maximum number of tokens in the strings of the 
   *                     input dataset (not used).
   * @param dictionarySize The size of the dictionary.
   * @tparam MatType The type of output matrix.
   */
```

File: src/mlpack/core/data/string_encoding_policies/bag_of_words_encoding_policy.hpp [hunk 22]
```diff
@@ -0,0 +1,141 @@
...
+#ifndef MLPACK_CORE_DATA_ENCODING_POLICIES_BAG_OF_WORDS_ENCODING_POLICY_HPP
+#define MLPACK_CORE_DATA_ENCODING_POLICIES_BAG_OF_WORDS_ENCODING_POLICY_HPP
+
+#include <mlpack/prereqs.hpp>
+#include <mlpack/core/data/string_encoding_policies/policy_traits.hpp>
+#include <mlpack/core/data/string_encoding.hpp>
+
+namespace mlpack {
+namespace data {
+/**
```
Comment (zoq) [inline src/mlpack/core/data/string_encoding_policies/bag_of_words_encoding_policy.hpp]:
```suggestion

/**
```

File: src/mlpack/core/data/string_encoding_policies/bag_of_words_encoding_policy.hpp [hunk 23]
```diff
@@ -0,0 +1,141 @@
...
+  * The function initializes the output matrix.
+  * Overloaded function to store result in vector<vector<OutputType>>
+  * 
+  * @param output Output matrix to store the encoded results.
+  * @param datasetSize The number of strings in the input dataset.
+  * @param maxNumTokens The maximum number of tokens in the strings of the 
+                        input dataset.
+  * @param dictionarySize The size of the dictionary (not used).
+  * @tparam OutputType The type of output vector.
+  */
```
Comment (zoq) [inline src/mlpack/core/data/string_encoding_policies/bag_of_words_encoding_policy.hpp]:
```suggestion
   * The function initializes the output matrix.
   * Overloaded function to store result in vector<vector<OutputType>>
   * 
   * @param output Output matrix to store the encoded results.
   * @param datasetSize The number of strings in the input dataset.
   * @param maxNumTokens The maximum number of tokens in the strings of the 
   *                     input dataset.
   * @param dictionarySize The size of the dictionary (not used).
   * @tparam OutputType The type of output vector.
   */
```

File: src/mlpack/core/data/string_encoding_policies/bag_of_words_encoding_policy.hpp [hunk 24]
```diff
@@ -0,0 +1,141 @@
...
+  /** 
+  * The function performs the bag of words encoding algorithm i.e. it writes
+  * the encoded token to the ouput.
+  *
+  * @param output Output matrix to store the encoded results (sp_mat or mat).
+  * @param value The encoded token.
+  * @param row The row number at which the encoding is performed.
+  * @param col The row token number at which the encoding is performed.
+  * @tparam MatType The type of output matrix.
+  */
```
Comment (zoq) [inline src/mlpack/core/data/string_encoding_policies/bag_of_words_encoding_policy.hpp]:
```suggestion
   * The function performs the bag of words encoding algorithm i.e. it writes
   * the encoded token to the output.
   *
   * @param output Output matrix to store the encoded results (sp_mat or mat).
   * @param value The encoded token.
   * @param row The row number at which the encoding is performed.
   * @param col The row token number at which the encoding is performed.
   * @tparam MatType The type of output matrix.
   */
```

File: src/mlpack/core/data/string_encoding_policies/bag_of_words_encoding_policy.hpp [hunk 25]
```diff
@@ -0,0 +1,141 @@
...
+  template<typename MatType>
+  static void Encode(MatType& output,
+                     size_t value,
+                     size_t row,
+                     size_t /*col*/)
+  {
+    // Important since Mapping of words,Dcitionary Encoding starts from 1,
```
Comment (zoq) [inline src/mlpack/core/data/string_encoding_policies/bag_of_words_encoding_policy.hpp]:
```suggestion
    // Important since Mapping of words, Dictionary Encoding starts from 1,
```

File: src/mlpack/core/data/string_encoding_policies/bag_of_words_encoding_policy.hpp [hunk 26]
```diff
@@ -0,0 +1,141 @@
...
+  * The function performs the bag of words encoding algorithm i.e. it writes
+  * the encoded token to the ouput.
+  * Overload function to accepted vector<vector<OutputType>> as output type.
+  *
+  * @param output Output matrix to store the encoded results.
+  * @param value The encoded token.
+  * @param row The row number at which the encoding is performed.
+  * @param col The row token number at which the encoding is performed.
+  * @tparam OutputType The type of output vector.
+  */
```
Comment (zoq) [inline src/mlpack/core/data/string_encoding_policies/bag_of_words_encoding_policy.hpp]:
```suggestion
   * The function performs the bag of words encoding algorithm i.e. it writes
   * the encoded token to the output.
   * Overload function to accepted vector<vector<OutputType>> as output type.
   *
   * @param output Output matrix to store the encoded results.
   * @param value The encoded token.
   * @param row The row number at which the encoding is performed.
   * @param col The row token number at which the encoding is performed.
   * @tparam OutputType The type of output vector.
   */
```

File: src/mlpack/core/data/string_encoding_policies/bag_of_words_encoding_policy.hpp [hunk 27]
```diff
@@ -0,0 +1,141 @@
...
+  template<typename OutputType>
+  static void Encode(std::vector<std::vector<OutputType> >& output,
+                     size_t value,
+                     size_t row,
+                     size_t /*col*/)
+  {
+    // Important since Mapping of words,Dcitionary Encoding starts from 1,
```
Comment (zoq) [inline src/mlpack/core/data/string_encoding_policies/bag_of_words_encoding_policy.hpp]:
```suggestion
    // Important since Mapping of words, Dictionary Encoding starts from 1,
```

File: src/mlpack/core/data/string_encoding_policies/bag_of_words_encoding_policy.hpp [hunk 28]
```diff
@@ -0,0 +1,141 @@
...
+    // Nothing to serialize.
+  }
+
+  /**
+  * Empty function, Important for tf-idf encoding policy
+  *
+  * @param row The row number at which the encoding is performed.
+  * @param numToken The count of token parsed till now.
+  * @param value The encoded token.
+  */
```
Comment (zoq) [inline src/mlpack/core/data/string_encoding_policies/bag_of_words_encoding_policy.hpp]:
```suggestion
   * Empty function, important for tf-idf encoding policy.
   *
   * @param row The row number at which the encoding is performed.
   * @param numToken The count of token parsed till now.
   * @param value The encoded token.
   */
```

File: src/mlpack/core/data/string_encoding_policies/tf_idf_encoding_policy.hpp [hunk 29]
```diff
@@ -0,0 +1,239 @@
...
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_CORE_DATA_ENCODING_POLICIES_TF_IDF_ENCODING_POLICY_HPP
+#define MLPACK_CORE_DATA_ENCODING_POLICIES_TF_IDF_ENCODING_POLICY_HPP
+
+#include <mlpack/prereqs.hpp>
+#include <mlpack/core/data/string_encoding_policies/policy_traits.hpp>
+#include <mlpack/core/data/string_encoding.hpp>
+namespace mlpack {
```
Comment (zoq) [inline src/mlpack/core/data/string_encoding_policies/tf_idf_encoding_policy.hpp]:
```suggestion

namespace mlpack {
```

File: src/mlpack/core/data/string_encoding_policies/tf_idf_encoding_policy.hpp [hunk 30]
```diff
@@ -0,0 +1,239 @@
...
+namespace data {
+/**
```
Comment (zoq) [inline src/mlpack/core/data/string_encoding_policies/tf_idf_encoding_policy.hpp]:
```suggestion

/**
```

File: src/mlpack/core/data/string_encoding_policies/tf_idf_encoding_policy.hpp [hunk 31]
```diff
@@ -0,0 +1,239 @@
...
+  /**
+  * Enum class used to identify the type of tf encoding
+  *
+  * Follwing are the defination of the types
+  * BINARY : binary weighting scheme (0,1)
+  * RAW_COUNT : raw count weighting scheme (count of token for every row)
+  * TERM_FREQUENCY : term frequency weighting scheme (count / length(row))
+  * SUBLINEAR_TF : logarthimic weighting scheme (log(tf) + 1) 
+  * 
+  */
```
Comment (zoq) [inline src/mlpack/core/data/string_encoding_policies/tf_idf_encoding_policy.hpp]:
```suggestion
   * Enum class used to identify the type of tf encoding
   *
   * Following are the the type definitions 
   * BINARY : binary weighting scheme (0,1)
   * RAW_COUNT : raw count weighting scheme (count of token for every row)
   * TERM_FREQUENCY : term frequency weighting scheme (count / length(row))
   * SUBLINEAR_TF : logarithmic weighting scheme (log(tf) + 1) 
   * 
   */
```

File: src/mlpack/core/data/string_encoding_policies/tf_idf_encoding_policy.hpp [hunk 32]
```diff
@@ -0,0 +1,239 @@
...
+  * and also the value for somoothIdf.
+  *
+  * @param tfType The type of term frequency, The avialbale option are
+  *     RAW_COUNT : The count of a specific token
+  *     BINARY : 1 if token occurs in document and 0 otherwise;
+  *     TERM_FREQUENCY :  Raw_count ÷ (number of words in document)
+  *     SUBLINEAR_TF : log(Raw_Count) + 1
+  *
+  * @param smoothIdf Used to indicate whether to use smooth idf or not.
+  */
```
Comment (zoq) [inline src/mlpack/core/data/string_encoding_policies/tf_idf_encoding_policy.hpp]:
```suggestion
   * A constructor for the class which is use to set the type of term frequency
   * and also the value for somoothIdf.
   *
   * @param tfType The type of term frequency, The available options are
   *     RAW_COUNT : The count of a specific token
   *     BINARY : 1 if token occurs in document and 0 otherwise;
   *     TERM_FREQUENCY :  Raw_count ÷ (number of words in document)
   *     SUBLINEAR_TF : log(Raw_Count) + 1
   *
   * @param smoothIdf Used to indicate whether to use smooth idf or not.
   */
```

File: src/mlpack/core/data/string_encoding_policies/tf_idf_encoding_policy.hpp [hunk 33]
```diff
@@ -0,0 +1,239 @@
...
+  /**
+  * The function initializes the output matrix.
+  *
+  * @param output Output matrix to store the encoded results (sp_mat or mat).
+  * @param datasetSize The number of strings in the input dataset.
+  * @param maxNumTokens The maximum number of tokens in the strings of the 
+                        input dataset.
+  * @param dictionarySize The size of the dictionary (not used).
+  * @tparam MatType The type of output matrix.
+  */
```
Comment (zoq) [inline src/mlpack/core/data/string_encoding_policies/tf_idf_encoding_policy.hpp]:
```suggestion
   * The function initializes the output matrix.
   *
   * @param output Output matrix to store the encoded results (sp_mat or mat).
   * @param datasetSize The number of strings in the input dataset.
   * @param maxNumTokens The maximum number of tokens in the strings of the 
   *                    input dataset.
   * @param dictionarySize The size of the dictionary (not used).
   * @tparam MatType The type of output matrix.
   */
```

**Comment by rcurtin** [comment]
@jeffin143 @lozhnikov this one's been on our radar for release for a little while; do you want to try and get it done by 3.3.0?  I'm happy to review it, but I'm not sure of the exact current status.  Looks like @zoq already got started reviewing it. :smile:  (I'm sure I've said it before, but I think adding string support to mlpack is really nice!  So I definitely support getting this in.  But if neither of you have bandwidth right now that's perfectly okay. :+1:)

**Comment by jeffin143** [comment]
> @jeffin143 @lozhnikov this one's been on our radar for release for a little while; do you want to try and get it done by 3.3.0? I'm happy to review it, but I'm not sure of the exact current status. Looks like @zoq already got started reviewing it.  (I'm sure I've said it before, but I think adding string support to mlpack is really nice! So I definitely support getting this in. But if neither of you have bandwidth right now that's perfectly okay. )

I am more than happy to do it , I will handle all the styling issues and spelling correction and other documentation issues, but since this pr is build on top of #2251 and hence we first need to merge those and then ultimately this

**Comment by rcurtin** [comment]
Ok, no problem---I'll try and review #2251 shortly.  We'll see if we can get it done in time but if not it's no issue, there are always future releases. :+1:

**Comment by lozhnikov** [comment]
@rcurtin Yes, this PR depends on #2251. I hope I'll be able to finish #2251 by Monday. So, I'll devote the next weekend to this PR.

**Comment by lozhnikov** [comment]
The PR is closed in favor of #2352.
